sequence_processing_pipeline/Pipeline.py

```diff
@@ -292,14 +317,220 @@ def generate_sample_information_files(self):
         results = []
 
-        for result in bioinformatics.to_dict('records'):
-            results.append({'project_name': result['Sample_Project'],
-                            'qiita_id': result['QiitaID']})
+        if self.mapping_file is not None:
+            sample_project_map = {}
+            for sample, project in zip(self.mapping_file.sample_name,
+                                       self.mapping_file.project_name):
+                if project not in sample_project_map:
+                    sample_project_map[project] = []
+                sample_project_map[project].append(sample)
+
+            # TODO: Replace 'something' w/proper Qiita ID
+            for project in sample_project_map:
+                qiita_id = project.split('_')[-1]
+                results.append(
+                    # assume project names end in qiita ids
+                    {'project_name': project, 'qiita_id': qiita_id})
+        else:
+            bioinformatics = self.sample_sheet.Bioinformatics
+            for result in bioinformatics.to_dict('records'):
+                results.append({'project_name': result['Sample_Project'],
+                                'qiita_id': result['QiitaID']})
 
         return results
+
+    @staticmethod
+    def is_mapping_file(mapping_file_path):
+        try:
+            Pipeline._validate_mapping_file(mapping_file_path)
+            return True
+        except PipelineError:
+            return False
+
+    @staticmethod
+    def _validate_mapping_file(mapping_file_path):
+        exp = {'barcode', 'library_construction_protocol', 'mastermix_lot',
+               'sample_plate', 'center_project_name', 'instrument_model',
+               'tm1000_8_tool', 'well_id', 'tm50_8_tool', 'well_description',
+               'run_prefix', 'run_date', 'center_name', 'tm300_8_tool',
+               'extraction_robot', 'experiment_design_description',
+               'platform', 'water_lot', 'project_name', 'pcr_primers',
+               'sequencing_meth', 'plating', 'orig_name', 'linker', 'runid',
+               'target_subfragment', 'primer', 'primer_plate', 'sample_name',
+               'run_center', 'primer_date', 'target_gene', 'processing_robot',
+               'extractionkit_lot'}
+        try:
+            df = pd.read_csv(mapping_file_path, delimiter='\t')
+
+            # if the two sets of headers are equal, then return the dataframe
+            # and no error message.
+            if len(exp - set(df.columns)) == 0:
+                return df
+
+            msg = 'missing columns: %s' % ', '.join(exp - set(df.columns))
+        except pd.errors.ParserError:
+            # ignore parser errors as they obviously prove this is not a
+            # valid mapping file.
+            msg = 'could not parse file "%s"' % mapping_file_path
+
+        raise PipelineError(msg)
+
+    def _generate_dummy_sample_sheet(self, first_read, last_read,
+                                     indexed_reads, dummy_sample_id):
```

Comment:
Also, just to confirm, why do we need _generate_dummy_sample_sheet and generate_dummy_sample_sheet?

Reply:
They were originally two different functions as they were tested on qiita-rc. I kept them separate instead of merging them since one is conceptually doing something different(error-checking and writing the file out). I could merge them I suppose?

Reply:
Nah, it's fine. Thank you.